### PR TITLE
Add usage documentation for Rich Text package

### PR DESCRIPTION
**Pull request by getdave**
## Description

I've been using the `@wordpress/rich-text` package a fair amount recently. It's super powerful and I really appreciate the work that's gone into this.

That said, to understand the package you must currently learn by either:

- reading the code
- reading the tests
- reviewing current example usages

Whilst it is possible to gain a working understanding of the fundamentals of the package in this way it is not the most efficient method. Given that this package is so critical and underpins much of what Gutenberg can do, this seems like something we should look to address. 

This PR attempts this by adding some basic `Usage` documentation. This covers

- the purpose of the package.
- the fundaments of the `RichTextValue` object.
- the concept of "selections" (including "collapsed" selections)

I would also like to cover the `replacements` property of the value object. I would do this but I still don't understand it so I'm going to need to rely on others to add this information.

## How has this been tested?

- Read the docs.
- Do they make sense?
- Is any element misleading?
- Is any part _inaccurate_? - specifically the code / prose-based examples.  
- Is my spelling and grammar ok?
- Was there anything missing?

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/rich-text/README.md [hunk 1]
```diff
@@ -12,6 +12,108 @@ npm install @wordpress/rich-text
 
 _This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for such language features and APIs, you should include [the polyfill shipped in `@wordpress/babel-preset-default`](https://github.com/WordPress/gutenberg/tree/HEAD/packages/babel-preset-default#polyfill) in your code._
 
+## Usage
+
+The Rich Text package is designed to facilitate the application and manipulation of rich formatting onto strings of plain text (e.g. bold, italic, links). This is facilitated via the usage of a `RichTextValue` value object (referred to from here on as `value`) which, by separating text from formatting, affords the ability to easily search and manipulate rich formats.
```
Comment (mkaz) [inline packages/rich-text/README.md]:
```suggestion
The Rich Text package applies rich formatting onto plain text strings (e.g. bold, italic, links). By using a `RichTextValue` value object (referred to from here on as `value`) separates the text from formatting, making it easier to search and manipulate rich formats.
```
I edited the intro, reworded here and there for clarity. I think still conveys the same info, let me know.

File: packages/rich-text/README.md [hunk 2]
```diff
@@ -12,6 +12,108 @@ npm install @wordpress/rich-text
...
+### The RichTextValue object
+
+The value object is comprised of the following:
+
+-   `text` - the string of text to which rich formats are to be applied.
+-   `formats` - a sparse array of the same length as `text` that is filled with [formats](https://developer.wordpress.org/block-editor/how-to-guides/format-api/) (e.g. `core/link`, `core/bold` etc.) at the positions where the text is formatted.
+-   `start` - an index in the `text` representing the _start_ of the currently active selection.
+-   `end` - an index in the `text` representing the _end_ of the currently active selection.
+
+You should not attempted to create your own `value` objects. Rather you should rely on the built in methods of the `@wordpress/rich-text` package to build these for you.
```
Comment (mkaz) [inline packages/rich-text/README.md]:
```suggestion
You should not attempt to create your own `value` objects. Rather you should rely on the built in methods of the `@wordpress/rich-text` package to build these for you.
```

File: packages/rich-text/README.md [hunk 3]
```diff
@@ -12,6 +12,108 @@ npm install @wordpress/rich-text
...
+
+Understanding how a value represents richly formatted text is very important so let's consider an example.
```
Comment (mkaz) [inline packages/rich-text/README.md]:
```suggestion
It is important to understand how a value represents richly formatted text. Here is an example to illustrate.
```

File: packages/rich-text/README.md [hunk 4]
```diff
@@ -12,6 +12,108 @@ npm install @wordpress/rich-text
...
+
+If `text` is formatted from position 2-5 in bold (`core/bold`) and from position 2-10 with a link (`core/link`), then you'll find:
```
Comment (mkaz) [inline packages/rich-text/README.md]:
```suggestion
If `text` is formatted from position 2-5 in bold (`core/bold`) and from position 2-8 with a link (`core/link`), then you'll find:
```

I think you use 8 in your example below.

**Review by mkaz — COMMENTED** [review]
Thanks for adding the examples, I've already learned more about the RichText package.

I've added a couple of suggestions on rewording in a few spots.

I was a little confused by the Collapsed selection, the first paragraph makes sense. However, the last sentence confused me on what I'm supposed to take away from it.

It looks like you are still working on it since there is a `TODO:` in the Replacements spot, or was that asking for others to add their input?

This is great so far and thanks so much for documenting and adding examples. It helps!

File: packages/rich-text/README.md (hunk 1 above)
Comment (getdave) [reply packages/rich-text/README.md]: I agree this was overly "wordy" 😆 

I want to convey that the package _aids_ in applying rich formatting. I'll tweak you're great suggested improvement.

File: packages/rich-text/README.md [hunk 5]
```diff
@@ -12,6 +12,108 @@ npm install @wordpress/rich-text
...
+
+#### Collapsed selections
+
+A collapsed selection is one where `start` and `end` values are identical. This happens when no characters are selected, but there is a caret present. This most often occurs when a user places the cursor/caret within a string of text but does not make a selected range.
+
+Given that it has no range identifiers, finding the currently selected portion of text from collapsed values can be challenging.
+
+### Replacements
+
+TODO: What are replacements? How do they work?
```
Comment (getdave) [inline packages/rich-text/README.md]: @gwwar Do you understand the concept of replacements? I feel like you do. Might you be able to help me fill in the blanks here?

File: packages/rich-text/README.md [hunk 6]
```diff
@@ -12,6 +12,108 @@ npm install @wordpress/rich-text
...
+
+If, as a user, I make a selection of the word `Hello` this would result in a value object with `start` and `end` as `0` and `5` respectively.
+
+In general, this is useful for knowing which portion of the text is selected. However, we need to consider that selections may also be "collapsed".
+
+#### Collapsed selections
+
+A collapsed selection is one where `start` and `end` values are identical. This happens when no characters are selected, but there is a caret present. This most often occurs when a user places the cursor/caret within a string of text but does not make a selected range.
+
+Given that it has no range identifiers, finding the currently selected portion of text from collapsed values can be challenging.
```
Comment (getdave) [inline packages/rich-text/README.md]: @mkaz I think this is what confused you. I agree it's completely unclear.

What I'm trying to convey is that as a collapsed selection has `start` and `end` that are _identical_ there is no "range" between them (unlike say, if you had `start: 1` and `end: 5`). The problem with this is that `start` and `end` are useful for identifying what portions of the `text` from the `RichTextValue` are _selected_ by the user. Unfortunately collapsed selections mean we loose this important info.

I believe Ella is looking to address this via modifications but it currently represents a challenge.

File: packages/rich-text/README.md (hunk 1 above)
Comment (getdave) [reply packages/rich-text/README.md]: Amended in https://github.com/WordPress/gutenberg/pull/35885/commits/c66a29f8f9230c9c98f19de2abf7da4fc0f2af62

File: packages/rich-text/README.md (hunk 6 above)
Comment (getdave) [reply packages/rich-text/README.md]: Addressed simplification of wording in https://github.com/WordPress/gutenberg/pull/35885/commits/28412590dbbcc458887e9920a9dffe4d38121a4e

File: packages/rich-text/README.md (hunk 5 above)
Comment (getdave) [reply packages/rich-text/README.md]: I've removed this section. Shall we tackle in a follow up?

**Review by mkaz — APPROVED** [review]
Thanks for updating, this is a good improvement to the RichText docs 👍

**Pull request merged**